### PR TITLE
Add ZStream#throttle (#1027)

### PR DESCRIPTION
**Pull request by vasilmkd**
Token bucket throttling implementation.

I would have liked to use ZSchedule to control the bucket refilling, but I can't seem to figure out how to create a schedule which adds an initial delay to the refilling effect, otherwise, the bucket is initialized with the configurable initial capacity and immediately refilled without any delay. Any help is appreciated.

This is my first open-source contribution. Looking forward to the comments.

### CODE REVIEW

**Comment by iravid** [comment]
Hey @vasilmkd, this is awesome! Thank you for this contribution, this is going to be super useful. I'll get to work on reviewing this tomorrow.

File: streams/shared/src/main/scala/zio/stream/ZSink.scala [hunk 1]
```diff
@@ -1063,4 +1094,50 @@ object ZSink extends ZSinkPlatformSpecific {
...
+
+    def refillFn(tokens: Long): Long = Math.min(bucketCapacity, tokens + refill)
+
+    def refillBucket(bucket: Ref[Long]) =
+      bucket.update(refillFn).unit.delay(refillInterval).repeat(ZSchedule.spaced(refillInterval))
+
+    for {
+      _      <- ZManaged.fromEffect(assertBucketConfiguration(bucketCapacity, initialTokens, refill))
+      bucket <- ZManaged.fromEffect(Ref.make(initialTokens))
+      sink   <- ZManaged.succeed(bucketSink(bucket))
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZSink.scala]: I think you could just `yield bucketSink(bucket)` below; should be more efficient

File: streams/shared/src/main/scala/zio/stream/ZSink.scala [hunk 2]
```diff
@@ -1063,4 +1094,50 @@ object ZSink extends ZSinkPlatformSpecific {
...
+      def extract(state: State) = UIO.succeed(state._2)
+    }
+
+    def refillFn(tokens: Long): Long = Math.min(bucketCapacity, tokens + refill)
+
+    def refillBucket(bucket: Ref[Long]) =
+      bucket.update(refillFn).unit.delay(refillInterval).repeat(ZSchedule.spaced(refillInterval))
+
+    for {
+      _      <- ZManaged.fromEffect(assertBucketConfiguration(bucketCapacity, initialTokens, refill))
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZSink.scala]: The `ZManaged.fromEffect(x)` calls can be replaced by `x.toManaged_`

**Comment by vasilmkd** [comment]
Awesome suggestions. I feel like I've overengineered the solution a bit. Thanks for the insightful comments.

File: streams/shared/src/main/scala/zio/stream/ZSink.scala (hunk 1 above)
Comment (vasilmkd) [reply streams/shared/src/main/scala/zio/stream/ZSink.scala]: Done.

File: streams/shared/src/main/scala/zio/stream/ZSink.scala (hunk 2 above)
Comment (vasilmkd) [reply streams/shared/src/main/scala/zio/stream/ZSink.scala]: Done.

**Comment by vasilmkd** [comment]
I'm interested to know if it's possible to integrate the test environment under testkit/ (TestClock in particular for this issue). It would be helpful in order to write a deterministic throttling test. However, I think it should be done in a different PR, as it will probably involve some build file manipulation.

File: streams/shared/src/main/scala/zio/stream/ZSink.scala [hunk 3]
```diff
@@ -1063,4 +1072,49 @@ object ZSink extends ZSinkPlatformSpecific {
...
+
+  /**
+   * Creates a sink which throttles input elements of type A according to the given bandwidth parameters
+   * using the token bucket algorithm. The weight of each element is determined by the `costFn` effectful
+   * function. Stream elements are mapped to `Option[A]`, and `None` denotes that a given element has
+   * been throttled.
+   */
+  final def throttle[R, E, A](units: Long, duration: Duration)(
+    costFn: A => ZIO[R, E, Long]
+  ): ZManaged[R with Clock, E, ZSink[R with Clock, E, Nothing, A, Option[A]]] = {
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZSink.scala]: So this doesn't need to be a `Managed` now right? The `Ref` and clock read can be done in `initial`.

File: streams/shared/src/main/scala/zio/stream/ZSink.scala [hunk 4]
```diff
@@ -1063,4 +1072,49 @@ object ZSink extends ZSinkPlatformSpecific {
...
       def extractPure(state: State): Either[Nothing, B]        = Right(b)
     }
+
+  /**
+   * Creates a sink which throttles input elements of type A according to the given bandwidth parameters
+   * using the token bucket algorithm. The weight of each element is determined by the `costFn` effectful
+   * function. Stream elements are mapped to `Option[A]`, and `None` denotes that a given element has
+   * been throttled.
+   */
+  final def throttle[R, E, A](units: Long, duration: Duration)(
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZSink.scala]: Let's call this `throttleEnforceM` and add a version `throttleEnforce` with `costFn: A => Long`

**Comment by iravid** [comment]
@vasilmkd Would be great if you could create follow-up tickets for things that should be deferred :-)

File: streams/shared/src/main/scala/zio/stream/ZSink.scala (hunk 3 above)
Comment (vasilmkd) [reply streams/shared/src/main/scala/zio/stream/ZSink.scala]: I forgot to change this. I wanted to send some code in before leaving for a festival. Working on this PR right now.

Comment (vasilmkd) [reply streams/shared/src/main/scala/zio/stream/ZSink.scala]: I have some trouble removing the `ZManaged` from the type signature. For each stream element, the `ZSink` gets reset (by calling `initial` state), which is why we need to keep a `Ref` as the state holder. Also, the ref initialization needs to happen exactly once, which means that we cannot make the ref in `initial`.

File: streams/shared/src/main/scala/zio/stream/ZSink.scala (hunk 4 above)
Comment (vasilmkd) [reply streams/shared/src/main/scala/zio/stream/ZSink.scala]: Done.

File: streams/shared/src/main/scala/zio/stream/ZSink.scala (hunk 3 above)
Comment (vasilmkd) [reply streams/shared/src/main/scala/zio/stream/ZSink.scala]: And since `initial` doesn't take any arguments, there is no way to "inject" the old state.

Comment (iravid) [reply streams/shared/src/main/scala/zio/stream/ZSink.scala]: Right, I see what you mean. Let's go ahead and merge this then.

**Pull request merged**